### PR TITLE
Break up the cookies classes

### PR DESCRIPTION
**Pull request by dwightwatson**
### Description of the Change
This was done in response to #1014.

* The various cookie classes in the cookie module are broken out into their own files.
* The `ChainedStore` module is just rolled into `Amber::Router::Cookies` where it is used.
* The `SerializedStore` and `JsonSerializer` are removed as there don't appear to be references to them.

I'm sure there is still work that can be done to cleanup this code, but I think this is a good first step.

### Alternate Designs
Leaving it in a single file as it was, or breaking it out to multiple files but keeping the old stuff around as well.

### Benefits
It'll be much easier to reason with and work with the cookie related code. It will also make it easier to attempt to DRY up any of the code that can be.

---

There appear to be test failures - but I'm not sure they are related. Would love a second opinion/point in the right direction. I'm having trouble replicating test failures locally - I don't know if it's something with my setup or something that I'm doing wrong.

### CODE REVIEW

**Comment by eliasjpr** [comment]
@dwightwatson the tests failures are related to these changes and it seems that the changes here might have broken the CLI. Try generating a new app with these changes and make sure the app runs successfully.

**Comment by eliasjpr** [comment]
I like the breakdown of the classes and moving in the right direction. Maybe abstract the duplication of methods

**Comment by dwightwatson** [comment]
I've merged master in after #1027 and the suite has gone back to green.

I've broken out an `AbstractStore` which matches the convention for session stores. It defines abstract methods `get` and `set` - which identified that the `PermanentStore` didn't actually conform to the protocol unlike the other stores - so I've updated it appropriately.

File: src/amber/router/cookies/abstract_store.cr [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    def [](name)
+      get(name)
+    end
+
+    def []=(name, value)
+      set(name, value)
+    end
+
+    abstract def get(name)
```
Comment (Sija) [inline src/amber/router/cookies/abstract_store.cr]: I'd use this opportunity to change it to `get?`. This way we have a free `#[]?` and a cheaper way to handle non-existent keys.

File: src/amber/router/cookies/store.cr [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    property host : String?
+    property secure : Bool = false
+
+    def initialize(@host = nil, @secret = Random::Secure.urlsafe_base64(32), @secure = false)
+      @cookies = {} of String => String
+      @set_cookies = {} of String => HTTP::Cookie
+      @delete_cookies = {} of String => HTTP::Cookie
+    end
+
+    def self.from_headers(headers)
```
Comment (Sija) [inline src/amber/router/cookies/store.cr]: This seems to be a constructor but instead it returns hash of `HTTP::Cookie`s.

File: src/amber/router/cookies/abstract_store.cr (hunk 1 above)
Comment (dwightwatson) [reply src/amber/router/cookies/abstract_store.cr]: If possible, I'd rather tackle that sort of improvement in another PR. The focus of this one was to break up the classes, remove dead code and have it all still work.

File: src/amber/router/cookies/store.cr (hunk 2 above)
Comment (dwightwatson) [reply src/amber/router/cookies/store.cr]: To be honest, I still find a lot of this code really confusing. Why is there a base Store class, but then we have other actual implementations of stores? I'd love to address further improvement here in other PR where I suspect we'd have some larger breaking changes.

Comment (Sija) [reply src/amber/router/cookies/store.cr]: Totally agree, that piece of the code screams for refactoring - in another PR that is.

**Pull request merged**